File: java/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example3/src/main/java/creational/singleton/example3/Singleton.java
package creational.singleton.example3;

// Thread-Safe Singleton using synchronized block,
// equivalent to C#'s lock(_lock) approach.
public class Singleton {

    private static Singleton instance;
    private static final Object lock = new Object();

    private Singleton() { }

    public static Singleton getInstance() {
        synchronized (lock) {
            if (instance == null) {
                instance = new Singleton();
            }
            return instance;
        }
    }
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example3/src/Singleton.ts
import { Mutex } from 'async-mutex';

// Node.js is single-threaded, but async operations can cause race conditions
// in concurrent scenarios. async-mutex is the Node.js equivalent of C#'s lock.
export class Singleton {
    private static _idCounter: number = 0;
    public readonly id: number;
    private static _instance: Singleton | null = null;
    private static readonly _mutex = new Mutex();

    private constructor() {
        Singleton._idCounter++;
        this.id = Singleton._idCounter;
    }

    static async getInstance(): Promise<Singleton> {
        const release = await Singleton._mutex.acquire();
        try {
            if (!Singleton._instance) {
                Singleton._instance = new Singleton();
            }
            return Singleton._instance;
        } finally {
            release();
        }
    }
}
